Enforce Maximum Warning Levels on Clang

tete17 opened the pull request:
This addresses #76 at the very least on Clang compiler.

There is a really good chance that this is also works for gcc but im uncapable of testing it, since i do not have access to the latest version of gcc. That is the reason the `add_compile_options` is not globall

tete17 committed:
Enable -Wall and -Werror on Clang

tete17 committed:
Enable -Wextra, -Wno-deprecated, -Werror, -pedantic and -pedantic-errors for Clang

tete17 committed:
Merge pull request #2 from arcosuc3m/master

Merge new commits

tete17 committed:
Merge branch 'Warnings_Clang'

# Conflicts:
#	CMakeLists.txt

tete17 committed:
Small Changes to enforce maximum warning levels on recent pulls on at the very least clang
Small bug in CMake file, that would not allow to compile after disabling Intel compiler

jdgarciauc3m commented on include/common/reduce_pattern.h:
```diff
@@ -41,7 +41,7 @@ class reduce_t {
   \param combine_op Combiner used for the reduction.
   */
-  reduce_t(int wsize, int offset, Identity id, Combiner && combine_op) :
+  reduce_t(size_t wsize, size_t offset, Identity id, Combiner && combine_op) :
```
In any place where `size_t` is used. It should be `std::size_t`.

However, I disagree with using unsigned for sizes and offsets. Please see CppCoreGuidelines ES.100 to ES.107

tete17 replied:
The issue relays on Line 68 and 78 of the same file. We are comparing an unsigned number (`.size()` aka `unsigned long`) with a signed number (aka `int`), clangs flags this as a warning. 
ES.100 recommends to avoid mixing them. It is true the second note kind of exclude this case, as the offset is a sort of difference of pointers but unless we plan on having negative offsets, a window size should always be positive.

jdgarciauc3m replied:
The current implementation of reduce_t (holding items in a vector) should not impact its interface.
However, I do agree that right now there is an issue in the current implementation with the comparison:

`items.size() >= window_size_`

The problem would arise when the number of elements in the reduction is greater than `std::numeric_limits<decltype(window_size_)>::max()`. This is only a theoretical issue because in that situation the reduction is useless and non-sense. 

Proposed resolution:
1.- Change the comparison to:
`static_cast<int>(items.size()) >= window_size_`

2.- Document a class invariant (in doxygen) stating that items.size() should always be below numeric_limits<int>::max(). Note that this invariant is true by design as `remaining_` max value is `offset_ - window_size_` and the difference is always to fit into an `int`

tete17 replied:
I think the best option is to use the `static_cast` solution

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -435,11 +435,11 @@ class parallel_execution_native {
 
   template <typename T, typename ... Others>
-  void do_pipeline(mpmc_queue<T> & in_q, mpmc_queue<T> & same_queue, Others &&... ops) const
+  void do_pipeline(mpmc_queue<T> &, mpmc_queue<T> &, Others &&...) const
```
OK. Unused, but this is going to complicate documentation.

Suggestion:
1.- Leave parameter names in declaration.
2.- Add a definition without parameter names.

tete17 replied:
Separated definition (with names for documentation) and definition without them, to avoid warnings

jdgarciauc3m replied:
OK

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -757,7 +757,7 @@ template <typename ... InputIterators, typename Identity,
     std::tuple<InputIterators...> firsts, 
     std::size_t sequence_size,
-    Identity && identity,
+    Identity &&,
```
We should probably open an issue to better understand why identity value is not being used in this version of map/reduce.

jdgarciauc3m replied:
I have opened issue #334 to handle this separately.

tete17 replied:
Fixed by changing the "child" sequential map_reduces to have the exact same input as the parent

jdgarciauc3m reviewed (changes requested):
Thanks for taking this one. Good job!

I have added some initial comments/feedback.

tete17 committed:
- Implemented feedback from pull request
- Fixed bug in map_reduce #334. Now every chunk does a regular sequential map_reduce with the exact same arguments as the parent.

jdgarciauc3m commented on CMakeLists.txt:
```diff
@@ -22,6 +22,7 @@ if ( ${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang" )
         message( FATAL_ERROR "Clang version " ${CMAKE_CXX_COMPILER_VERSION} " not supported. Upgrade to 3.9 or above.")
     else()
         message( STATUS "C++ Compiler is Clang" )
+        add_compile_options(-Wall -Wextra -Wno-deprecated -Werror -pedantic -pedantic-errors)
```
We should not have -Wno-deprecated that would allow using deprecated language features.

jdgarciauc3m commented on include/common/reduce_pattern.h:
```diff
@@ -65,7 +65,7 @@ class reduce_t {
   \brief Check if a reduction can be performed.
   */
```
We should add a doxygen precondtion

`\pre items.size() < static_cast<int>(std::numeric_limits<int>::max())`

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -757,7 +756,7 @@ template <typename ... InputIterators, typename Identity,
     std::tuple<InputIterators...> firsts, 
     std::size_t sequence_size,
-    Identity && identity,
+    Identity && identity_,
```
trailing underscores in identifiers is reserved in GrPPI for class data members.

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -766,7 +765,7 @@ auto parallel_execution_native::map_reduce(
   auto process_chunk = [&](auto f, std::size_t sz, std::size_t id) {
     partial_results[id] = seq.map_reduce(f, sz,
-        std::forward<Identity>(partial_results[id]), 
+        std::forward<Identity>(identity_),
```
Why?

tete17 replied:
The idea is that now every "child" map_reduce uses the same arguments as the parent.
This changes a bit the design. Let me know if you think we should only use the identity on the first map_reduce child and use the first element of every chunk as a sort of identity, the same way it is done in the last step of this algorithm to put in common all results form each parallized thread.

jdgarciauc3m replied:
I see your approach.
We need that this is double checked by @Drioastorga to make sure we are not forgetting any corner case.

jdgarciauc3m replied:
Note that this is related to issue #334

Drioastorga replied:
Just a comment. This change is only valid if the identity value introduced is the identity value for the operator. Otherwise the value introduced as identity will be part of each partial result. For instance, if the combine operator is a plus operator, and the user inroduces "2" as identity value, these value will be added number_of_ threads times in the final result.

Therefore, we have two choices:
- To perform this modification and document this behaviour. Thus, only when the value is an actual identity value the results will be correct in parallel versions.
- To not modify and maintain the original behaviour.

jdgarciauc3m replied:
We have discussed this and we concluded that we SHOULD accept this change.

tete17 replied:
Revert back to the original in this pull, except removing the name of the parameter. We should recover it once we fix this issue.

jdgarciauc3m commented on include/tbb/parallel_execution_tbb.h:
```diff
@@ -559,7 +559,7 @@ template <typename ... InputIterators, typename Identity,
     std::tuple<InputIterators...> firsts,
     std::size_t sequence_size,
-    Identity && identity,
+    Identity && identity_,
```
Trailing underscores is only for data members.

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -1164,7 +1150,7 @@ void parallel_execution_native::do_pipeline(Queue & input_queue,
     get_output_queue<output_item_type>(other_ops...);
   
-  auto context_task = [&](int nt) {
+  auto context_task = [&](int) {
```
Make this lambda not taking any argument

`auto context_tasks = [&]() {`

jdgarciauc3m commented on include/native/parallel_execution_native.h:
```diff
@@ -1464,6 +1446,11 @@ void parallel_execution_native::do_pipeline_nested(
 }
 
+template<typename T, typename... Others>
+void parallel_execution_native::do_pipeline(mpmc_queue<T> &, mpmc_queue<T> &, Others &&...) const
```
Where does this one come from?
Wasn't it in the class delcaration?

tete17 replied:
It was in the class declaration but due to the parameters not being used, and for documentation purpose we had to move it outside so that the parameters inside the class could have names and not raise warning by the compilers

jdgarciauc3m commented on include/omp/parallel_execution_omp.h:
```diff
@@ -338,11 +338,11 @@ class parallel_execution_omp {
 
   template <typename T, typename ... Others>
-  void do_pipeline(mpmc_queue<T> & in_q, mpmc_queue<T> & same_queue, Others &&... ops) const
+  void do_pipeline(mpmc_queue<T> &, mpmc_queue<T> &, Others &&...) const
```
Do not remove parameter names from in-class declaration. Just move implementation outside.

jdgarciauc3m commented on include/omp/parallel_execution_omp.h:
```diff
@@ -661,7 +661,7 @@ template <typename ... InputIterators, typename Identity,
     std::tuple<InputIterators...> firsts,
     std::size_t sequence_size,
-    Identity && identity,
+    Identity && identity_,
```
Conding stanadard: Trailing underscores are for data members.

jdgarciauc3m commented on unit_tests/farm_pipeline_composition.cpp:
```diff
@@ -50,8 +50,8 @@ class farm_pipeline_test : public ::testing::Test {
   // entry counter
-  int idx_in = 0;
-  int idx_out = 0;
+  size_t idx_in = 0;
```
Why?

tete17 replied:
We have comparision with a .size() in lines 170, 216, 259 and 303.
Since it is a unit test, and indexes should always be positive i think is best to change them to size_t instead of using a static cast every time we compare them

jdgarciauc3m replied:
I'm not 100% convinced.

But OK. Let's do it for now.

tete17 replied:
Size comparison in line 170. Similar issue pointed out previously. I think the size_t solution is best than a static_cast since it is just a unit test and this code is under our control and not something the user will be "using"

jdgarciauc3m reviewed (changes requested):
More comments to address

jdgarciauc3m reviewed (changes requested):
There are several issues to be addressed here.

Besides, the identity issue should be handled in a separate pull request.

tete17 committed:
More feedback of Pull.
Removed proposed solution of issue #334 to be implemented on another pull

tete17 committed:
Merge branch 'master' into master

jdgarciauc3m reviewed (changes requested):
All code seems OK now.
However, I miss the compile options setting for warnings.
It seems it was lost in last merges. Can you please restore that in CMake?

jdgarciauc3m reviewed (approved):
I am approving the PR. 
However, the issue will not be closed until the warnings are enabled.
